src/pcap/agent.go

```diff
@@ -0,0 +1,367 @@
+package pcap
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/pcap"
+	"github.com/google/uuid"
+	"go.uber.org/zap"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+)
+
+// Agent is the central struct to which the handlers are attached.
+type Agent struct {
+	// done is used to gracefully shut down the agent, all ongoing streams terminate
+	// whenever this channel is closed.
+	done chan struct{}
+	// wg tracks any running streams.
+	// TODO: expose as metric?
+	wg sync.WaitGroup
+	// log carries the logger all session loggers are derived from.
+	log     *zap.Logger
+	bufConf BufferConf
+
+	UnimplementedAgentServer
+}
+
+// NewAgent creates a new ready-to-use agent. If the given logger is nil zap.L will
+// be used.
+func NewAgent(log *zap.Logger, bufConf BufferConf) (*Agent, error) {
+	if log == nil {
+		log = zap.L()
+	}
+	return &Agent{
+		done:    make(chan struct{}),
+		log:     log,
+		bufConf: bufConf,
+	}, nil
+}
+
+// Stop the server. This will gracefully stop any captures that are currently running
+// by closing Agent.done. Further calls to Stop have no effect.
+func (a *Agent) Stop() {
+	select {
+	case <-a.done:
+		// if the channel is already closed, we do nothing
+	default:
+		// otherwise the channel is still open and we close it
+		close(a.done)
+	}
+}
+
+// Wait for all open streams to terminate.
+func (a *Agent) Wait() {
+	a.wg.Wait()
+}
+
+// draining returns true after Agent.Stop has been called.
+func (a *Agent) draining() bool {
+	select {
+	case <-a.done:
+		// we only get here if the channel is closed since it is never written to
+		return true
+	default:
+		// channel is still open
+		return false
+	}
+}
+
+// Status handler for the pcap-agent. See AgentServer.Status documentation for details.
+func (a *Agent) Status(_ context.Context, _ *StatusRequest) (*StatusResponse, error) {
+	s := &StatusResponse{
+		CompatibilityLevel: CompatibilityLevel,
+		Health:             Health_UP,
+		Status:             "ok",
+	}
+
+	if a.draining() {
+		s.Health = Health_DRAINING
+		s.Status = "agent is shutting down"
+	}
+
+	return s, nil
+}
+
+// Capture handler for the pcap-agent. See AgentServer.Capture documentation for details.
+func (a *Agent) Capture(stream Agent_CaptureServer) (err error) {
+	a.wg.Add(1)
+	defer a.wg.Done()
+
+	log := a.log.With(zap.String("handler", "capture"))
+	defer func() {
+		if err != nil {
+			log.Error("capture ended unsuccessfully", zap.Error(err))
+		}
+	}()
+
+	ctx, cancel := WithCancelCause(stream.Context())
+	defer cancel(nil)
+
+	setVcapID(ctx, log)
+
+	if a.draining() {
+		return errorf(codes.Unavailable, "agent is draining")
+	}
```

Comment:
this should also be done by middleware, e.g. a `checkDraining` handler. `Capture` handler should only focus on capture logic.